tests/Mono.Android-Tests/Xamarin.Android.Net/AndroidMessageHandlerTests.cs

```diff
@@ -85,15 +76,51 @@ public async Task ServerCertificateCustomValidationCallback_ApprovesRequestWithI
 			};
 
 			var client = new HttpClient (handler);
+			await client.GetStringAsync ("https://self-signed.badssl.com/");
 
-			try {
-				await client.GetStringAsync ("https://self-signed.badssl.com/");
-			} catch (Exception e) {
-				exception = e;
-			}
+			Assert.IsTrue (callbackHasBeenCalled, "custom validation callback hasn't been called");
+		}
+
+		[Test]
+		public async Task NoServerCertificateCustomValidationCallback_ThrowsWhenThereIsCertificateHostnameMismatch ()
+		{
+			var handler = new AndroidMessageHandler ();
+			var client = new HttpClient (handler);
+
+			await AssertRejectsRemoteCertificate (() => client.GetStringAsync ("https://wrong.host.badssl.com/"));
+		}
+
+		[Test]
+		public async Task ServerCertificateCustomValidationCallback_IgnoresCertificateHostnameMismatch ()
+		{
+			bool callbackHasBeenCalled = false;
+			SslPolicyErrors reportedErrors = SslPolicyErrors.None;
+
+			var handler = new AndroidMessageHandler {
+				ServerCertificateCustomValidationCallback = (request, cert, chain, errors) => {
+					callbackHasBeenCalled = true;
+					reportedErrors = errors;
+					return true;
+				}
+			};
+
+			var client = new HttpClient (handler);
+			await client.GetStringAsync ("https://wrong.host.badssl.com/");
 
 			Assert.IsTrue (callbackHasBeenCalled, "custom validation callback hasn't been called");
-			Assert.IsNull (exception, $"an exception was thrown: {exception}");
+			Assert.AreEqual (SslPolicyErrors.RemoteCertificateNameMismatch, reportedErrors & SslPolicyErrors.RemoteCertificateNameMismatch);
+		}
+
+		private async Task AssertRejectsRemoteCertificate (Func<Task> makeRequest)
+		{
+			try {
+				await makeRequest();
+			} catch (Java.IO.IOException) {
+				// the Java.IO.IOException is thrown when the TLS connection cannot be established
```

Comment:
If a TLS connection can't be established, then presumably an `HttpRequestException` should be thrown, with an *inner exception* of `Java.IO.IOException`.

Reply:
I agree that we should actually throw `HttpRequestException` but that will involve modifying a [different part of AndroidMessageHandler](https://github.com/xamarin/xamarin-android/blob/main/src/Mono.Android/Xamarin.Android.Net/AndroidMessageHandler.cs#L418:L426) and consider how this will affect customers who rely on this behavior in their existing codebases. I believe that's outside of the scope of this PR.